keps/sig-network/0015-20180614-SCTP-support.md

```diff
@@ -0,0 +1,258 @@
+
+#### SCTP in NetworkPolicy
+The Kubernetes API modification for the NetworkPolicy is obvious.
+
+In order to utilize the new protocol value the network controller must support it.
+
+#### Interworking with applications that use a user space SCTP stack
+A userspace SCTP stack implementation cannot work together with the SCTP kernel module (lksctp) on the same node. That is, the loading of the SCTP kernel module must be avoided on nodes where such applications that use userspace SCTP stack are planned to be run. The problem comes with the introduction of the SCTP protocol option for Services with Virtual IP (the "type" of the Service is ClusterIP or NodePort): once such a service is created the relevant port reservation logic kicks-in on every node in the cluster, it starts listening on the port, and as a consequence it loads the SCTP kernel module on every nodes. It immediately ruins the connectivity of the userspace SCTP applications on those nodes. 
+
+The same interworking problem stands for Services with "externalIP" defined.
+
+NOTE! It is not a new interworking problem between the userspace SCTP stack implementations and the SCTP kernel module. It is a known phenomenon. The userpace SCTP stack creates raw sockets with IPPROTO_SCTP. As it is clearly highlighted in the [documentation of raw sockets][]:
+>Raw sockets may tap all IP protocols in Linux, even protocols like ICMP or TCP which have a protocol module in the kernel.  In this case, the packets are passed to both the kernel module and the raw socket(s).
+
+I.e. it is the normal function of the [kernel][], that it sends the incoming packet to both sides: the raw socket and relevant the kernel module. In this case the kernel module will handle those packets that are destined to the raw socket as Out of the blue (OOTB) packets  according to the rules defined in the [RFC4960][].
+
+In order to resolve this problem the solution has been to dedicate nodes to userspace SCTP applications, and to ensure that on those nodes the SCTP kernel module is not loaded.
+
+For this reason the main task here is to provide the same isolation possibility: i.e. to provide the option to dedicate some nodes to userspace SCTP applications and ensure that the actions performed by Kubernetes do not load the SCTP kernel modules on those dedicated nodes.
+
+As we can easily see, it is pretty easy to separate application pods that use a userspace SCTP stack from those application pods that use the kernel space SCTP stack: the usual nodeselector label based mechanism,  or taints are there for this very purpose. 
+
+The real challenge here is to ensure that when an SCTP Service is created in a Kubernetes cluster the Kubernetes logic does not create listening SCTP sockets on those nodes that are dedicated for the applications that use  userspace SCTP stack - because such an action would trigger the loading of the kernel module.
```

Comment:
It's not clear to me.

Is this proposal saying that Kubelet and kube-proxy will not do the usual "open and hold" operation for SCTP ports?

Is that sufficient?  If I have a cluster with an SCTP service, that triggers loading iptables modules for sctp, but does not listen() on the port - will userspace SCTP be OK?

Reply:
> Is this proposal saying that Kubelet and kube-proxy will not do the usual "open and hold" operation for SCTP ports?

Yeah, on nodes that were marked userspace-sctp. Which would mean that HostPort/ExternalIP SCTP services wouldn't work on userspace-sctp nodes, and NodePort SCTP services would not be reachable on userspace-sctp nodes.

But that weirdness would be limited to clusters that used userspace sctp; in clusters with no userspace sctp nodes, SCTP would behave just like everything else.

> If I have a cluster with an SCTP service, that triggers loading iptables modules for sctp, but does not listen() on the port - will userspace SCTP be OK?

It looks like the kernel has some code to parse SCTP packets even without the `sctp` module being loaded, and you can add rules with matches like `-p sctp --dport 80` without needing the kernel sctp module. (It will cause `xt_sctp` to be loaded, but that's just part of iptables, and wouldn't interfere with userspace SCTP processing.)

It is possible that some iptables extension rules (NAT?) might cause `sctp` to be loaded.

Reply:
I'm not very excited by formalizing userspace SCTP as a concept in Kubernetes based on labels.  In general, we don't want to use annotations or labels for config, when we have real config (even if kube-proxy is behind the curve wrt component config).

We could add a flag/component-config param here, and leave labels to users.  I would be OK with that.

Reply:
Wrt SCTP NAT: according to our observations, iptables based NAT does not cause any harm. As @KomorkinMikhail wrote above in this PR, when they implemented SCTP related patches into k8s they simply skipped those "openlocalport" parts for SCTP, and they could use it with NodePort and ExternalIP as well.

Reply:
OK.  This wasn't clear.  The main points, if I understand:

1) skip the opening/holding of the port in both kubelet and kube-proxy IFF SCTP
2) suggest that user segregate their user-space SCTP work and their kernel-space SCTP work by node labels, but the details of that are left to the user

Reply:
Yepp, my English.... :/
Exactly, that was the plan.